src/charm.py

```diff
@@ -1,427 +1,127 @@
-            obj_storage = _get_obj_storage(interfaces)
-            secrets = self._define_secrets(obj_storage=obj_storage, mysql=mysql)
+        self.logger = logging.getLogger(__name__)
+        self._port = self.model.config["mlflow_port"]
+        self._container_name = "mlflow-server"
+        self._container = self.unit.get_container(self._container_name)
 
-            default_artifact_root = self._validate_default_s3_bucket(obj_storage)
-
-            self._configure_mesh(interfaces)
-        except CheckFailedError as check_failed:
-            self.model.unit.status = check_failed.status
-            self.model.unit.message = check_failed.msg
-            return
+        self.framework.observe(self.on.upgrade_charm, self._on_event)
+        self.framework.observe(self.on.config_changed, self._on_event)
+        self.framework.observe(self.on.mlflow_server_pebble_ready, self._on_pebble_ready)
 
-        self.model.unit.status = MaintenanceStatus("Setting pod spec")
+        self._create_service()
 
-        config = self.model.config
-
-        pod_spec_services = self._get_pod_spec_services(config)
-
-        self.model.pod.set_spec(
-            {
-                "version": 3,
-                "containers": [
-                    {
-                        "name": "mlflow",
-                        "imageDetails": image_details,
-                        "ports": [{"name": "http", "containerPort": config["mlflow_port"]}],
-                        "args": [
-                            "--host",
-                            "0.0.0.0",
-                            "--backend-store-uri",
-                            "$(MLFLOW_TRACKING_URI)",
-                            "--default-artifact-root",
-                            f"s3://{default_artifact_root}/",
-                            "--expose-prometheus",
-                            "{}".format(PROMETHEUS_PATH),
-                        ],
-                        "envConfig": {
-                            "db-secret": {"secret": {"name": f"{self.charm_name}-db-secret"}},
-                            "aws-secret": {"secret": {"name": f"{self.charm_name}-minio-secret"}},
-                            "AWS_DEFAULT_REGION": "us-east-1",
-                            "MLFLOW_S3_ENDPOINT_URL": _gen_obj_storage_endpoint_url(obj_storage),
-                        },
-                    }
-                ],
-                "kubernetesResources": {
-                    "secrets": secrets,
-                    "services": pod_spec_services,
-                },
-            },
-        )
-        self.model.unit.status = ActiveStatus()
+    @property
+    def container(self):
+        """Return container."""
+        return self._container
 
-    def _get_pod_spec_services(self, config):
-        """Returns service list for pod spec based on enabled service flags."""
-        pod_spec_services = []
+    def _create_service(self):
         if self.config["enable_mlflow_nodeport"]:
```

Comment:
This if/else might be more readable if it only set the stuff that differs between the two branches, rather than have it all duplicated.  I think atm it is somehing like:
```python
if True:
  x = "true branch"
  y = "true branch"
  do_something(
    common_setting,
    x,
    y
  )
else:
  x = "false branch"
  y = "false branch"
  do_something(
    common_setting,
    x,
    y
  )
```
which makes the reader need to focus to understand what is different.  What about writing it like:
```python
if True:
  x = "true branch"
  y = "true branch"
else:
  x = "false branch"
  y = "false branch"
do_something(
  common_setting,
  x,
  y
)
```
that way the common parts are easy to spot

Reply:
Solved thanks :)